app/cni/pkg/cni/kubernetes_linux.go

```diff
@@ -2,41 +2,108 @@ package cni
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: conf.Kubernetes.Kubeconfig},
 		&clientcmd.ConfigOverrides{},
 	).ClientConfig()
 	if err != nil {
-		log.Error(err, "failed setting up kubernetes client with kubeconfig", "kubeconfig", kubeconfig)
+		logger.Error(err, "failed setting up kubernetes client with kubeconfig")
 		return nil, err
 	}
 
-	log.V(1).Info("set up kubernetes client with kubeconfig", "kubeconfig", kubeconfig, "config", config)
+	logger.V(1).Info("set up kubernetes client with kubeconfig", "config", config)
 
 	return kubernetes.NewForConfig(config)
 }
 
-// getK8sPodInfo returns information of a POD
-func getKubePodInfo(ctx context.Context, client *kubernetes.Clientset, podName, podNamespace string) (int, map[string]struct{}, map[string]string, error) {
-	pod, err := client.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
-	log.V(1).Info("pod info", "pod", pod)
+func getAndValidatePodAnnotations(
+	ctx context.Context,
+	logger logr.Logger,
+	conf *PluginConf,
+	k8sArgs K8sArgs,
+) (map[string]string, bool, error) {
+	name := string(k8sArgs.K8S_POD_NAME)
+	namespace := string(k8sArgs.K8S_POD_NAMESPACE)
+
+	if namespace == "" || name == "" {
+		return nil, true, errors.New("pod namespace or pod name is empty")
+	}
+
+	if slices.Contains(conf.Kubernetes.ExcludeNamespaces, namespace) {
+		return nil, true, errors.New("namespace is in the exclude list")
+	}
+
+	client, err := newKubeClient(logger, *conf)
 	if err != nil {
-		log.Error(err, "can't get pod info")
-		return 0, nil, nil, err
+		return nil, false, errors.Wrap(err, "failed to create Kubernetes client")
+	}
+
+	var pod *corev1.Pod
+	if err := retry.Do(
+		ctx,
+		retry.WithMaxRetries(podRetrievalMaxRetries, retry.NewConstant(podRetrievalInterval)), // backoff
+		func(ctx context.Context) error {
+			pod, err = client.CoreV1().Pods(namespace).Get(ctx, name, metav1.GetOptions{})
+			return retry.RetryableError(err)
+		},
+	); err != nil {
+		return nil, false, errors.Wrap(err, "failed to retrieve pod data from Kubernetes API")
+	}
+
+	logger.V(1).Info(
+		"retrieved pod data",
+		"name", pod.Name,
+		"namespace", pod.Namespace,
+		"annotations", pod.Annotations,
+		"containersCount", len(pod.Spec.Containers),
+		"initContainersCount", len(pod.Spec.InitContainers),
+	)
+
+	containers := make(map[string]struct{}, len(pod.Spec.Containers))
+	for _, container := range pod.Spec.Containers {
+		containers[container.Name] = struct{}{}
+	}
+
+	initContainers := make(map[string]struct{}, len(pod.Spec.InitContainers))
+	for _, container := range pod.Spec.InitContainers {
+		initContainers[container.Name] = struct{}{}
+	}
+
+	if _, ok := initContainers[k8s_util.KumaInitContainerName]; ok {
+		return nil, true, errors.New("pod already injected with kuma-init container")
+	}
+
+	if len(containers) < 2 {
+		return nil, true, errors.New("pod requires at least two containers; kuma-sidecar container is missing")
```

Comment:
what about native sidecars? In this case it would end up in init containers list

Reply:
addressed